### DISCUSSION TITLE
Realistic Contact Shadows

### DISCUSSION THREAD

**Discussion by dhruvadhia1** [google/model-viewer#2261] [Ideas]
It would be interesting to achieve contact shadows to provide better realism to the objects. 

Here are some references. 

https://threejs.org/examples/webgl_shadow_contact.html

https://forum.babylonjs.com/t/realistic-contact-shadow-ambient-occlusion/14733

**Comment by elalish** [comment]
We do have shadows if you turn them on. Do you have specific changes you would like to see?

**Reply by dhruvadhia1** [reply]
Shadows are mostly on the ground plane, do we have contact shadows one object onto another?

**Reply by elalish** [reply]
No, not currently as they add considerable rendering expense, lowering frame rate. For static models the better approach is to bake the self-shadows into the AO map. Are you concerned specifically with animated models?

**Reply by dhruvadhia1** [reply]
Yes for moving objects. Baking into AO map is definitely a solution for now.

**Reply by slightHYP** [reply]
Hi, is there any plan to add this feature?

I completley understand that this will affect the rendering performance, but our 3d content creation team would like to add some animations to the models introducing the casting shadow issue.

At the moment my idea is to apply the AO map at the end of the animation, but this addition/switch could be not so soft and in term of UX not so smooth.

I found this https://github.com/google/model-viewer/discussions/2278

So I thought to add some fade-in/out or cross effect of a texture map (not so realistic anyway, but could be nicer), but I don't know if this is possible? It could be very complicated.



What do you think, are there any better approaches? Thanks

**Reply by elalish** [reply]
I think that's out of scope for us; there are just too many options and considerations. You can do this pretty easily by forking us and enabling them in the three.js layer. Can you post some links or videos of what you're trying to achieve and the problems you're having so I can better understand?

**Reply by slightHYP** [reply]
Hi, I will try to describe our use case.

Our 3D team created a sofa with chest within an animation to open it (something like [this](https://www.loutletshop.com/712-large_default/divano-letto-contenitore-murphy.jpg))

We included it in model-viewer within an HDR environment image and the animation is activable by a button of the UI. Everything works well, smoth and the quality is fine.

The issue highlighted by the 3D team is that when the sofa is opened we should add some extra shadows (contact shadows) produced by the sofa itself on some parts like the back and the armrests.

I explained them that we are not in a real-time game engine and some features are not available for obvious reasons of performance and they understood this point, but at the same time I would like to find a workaround to satisfy the request.

My idea is to apply a new AO map during the animation (or at least at the end) within a fade-in transition.

I found [this](https://jsfiddle.net/jq8c4hpL/) to explain the transition effect. Using something like this on the AO map, but I'm not a shaders expert at all and I don't know if this could be supported in model-viewer and, in case, the effort to accomplish this.

Thank you for your help

**Reply by elalish** [reply]
Yeah, we don't have support for fading a texture. You can switch it, but the pop would probably be distracting. A video texture might be an option, but it feels like overkill and we don't support that either yet.

**Reply by slightHYP** [reply]
Thank you!
